dbms/src/Columns/ColumnString.cpp

```diff
@@ -596,116 +919,168 @@ void ColumnString::serializeToPosForColumnArrayImpl(
+            {
+                UInt32 str_size = sizeAt(j);
+                const void * src = &chars[offsetAt(j)];
+                auto sort_key
+                    = derived_collator->sortKey(reinterpret_cast<const char *>(src), str_size - 1, *sort_key_container);
+                str_size = sort_key.size;
+                src = sort_key.data;
+
+                tiflash_compiler_builtin_memcpy(pos[i], &str_size, sizeof(UInt32));
+                pos[i] += sizeof(UInt32);
+                inline_memcpy(pos[i], src, str_size);
+                pos[i] += str_size;
+            }
         }
-        for (size_t j = array_offsets[start + i - 1]; j < array_offsets[start + i]; ++j)
+    }
+    else
+    {
+        for (size_t i = 0; i < length; ++i)
         {
-            UInt32 str_size = sizeAt(j);
-            tiflash_compiler_builtin_memcpy(pos[i], &str_size, sizeof(UInt32));
-            pos[i] += sizeof(UInt32);
+            if constexpr (has_null)
+            {
+                if (pos[i] == nullptr)
+                    continue;
+            }
+            for (size_t j = array_offsets[start + i - 1]; j < array_offsets[start + i]; ++j)
+            {
+                UInt32 str_size = sizeAt(j);
+                tiflash_compiler_builtin_memcpy(pos[i], &str_size, sizeof(UInt32));
+                pos[i] += sizeof(UInt32);
+            }
+            size_t strs_size = offsetAt(array_offsets[start + i]) - offsetAt(array_offsets[start + i - 1]);
+            inline_memcpy(pos[i], &chars[offsetAt(array_offsets[start + i - 1])], strs_size);
+            pos[i] += strs_size;
         }
-        size_t strs_size = offsetAt(array_offsets[start + i]) - offsetAt(array_offsets[start + i - 1]);
-        inline_memcpy(pos[i], &chars[offsetAt(array_offsets[start + i - 1])], strs_size);
-        pos[i] += strs_size;
     }
 }
 
-void ColumnString::deserializeAndInsertFromPos(PaddedPODArray<char *> & pos, bool use_nt_align_buffer [[maybe_unused]])
+void ColumnString::deserializeForCmpAndInsertFromPos(
+    PaddedPODArray<char *> & pos,
+    bool use_nt_align_buffer,
+    const TiDB::TiDBCollatorPtr & collator)
+{
+    if likely (collator != nullptr)
+        deserializeAndInsertFromPosImpl<true>(pos, use_nt_align_buffer);
+    else
+        deserializeAndInsertFromPosImpl<false>(pos, use_nt_align_buffer);
+}
+
+void ColumnString::deserializeAndInsertFromPos(PaddedPODArray<char *> & pos, bool use_nt_align_buffer)
+{
+    deserializeAndInsertFromPosImpl<false>(pos, use_nt_align_buffer);
+}
+
+template <bool add_terminating_zero>
+void ColumnString::deserializeAndInsertFromPosImpl(
+    PaddedPODArray<char *> & pos,
+    bool use_nt_align_buffer [[maybe_unused]])
 {
     size_t prev_size = offsets.size();
     size_t char_size = chars.size();
     size_t size = pos.size();
 
 #ifdef TIFLASH_ENABLE_AVX_SUPPORT
-    if (use_nt_align_buffer)
+    if constexpr (!add_terminating_zero)
```

Comment:
How about adding a terminating zero in `serializeToPos`? If so, this branch parameter can be removed and align_buffer can be used as original version.